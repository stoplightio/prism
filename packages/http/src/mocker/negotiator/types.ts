import { IHttpOperationConfig } from '@stoplight/prism-http';
import { INodeExample, INodeExternalExample, ISchema } from '@stoplight/types';

export interface IHttpNegotiationResult {
  code: string;
  mediaType: string;
  example?: INodeExample | INodeExternalExample;
  schema?: ISchema;
}

export type NegotiationOptions = IHttpOperationConfig;

export type NegotiatePartialOptions = {
  code: string;
  dynamic: boolean;
  exampleKey?: string;
};
