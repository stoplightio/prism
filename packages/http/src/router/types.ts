import { IHttpOperationEx } from '../types';

export type Nullable<T> = T | null;

export interface IMatch {
  resource: IHttpOperationEx;
  pathMatch: MatchType;
  methodMatch: MatchType;
  serverMatch?: Nullable<MatchType>;
}

export enum MatchType {
  CONCRETE = 'concrete',
  TEMPLATED = 'templated',
  NOMATCH = 'no-match',
}
