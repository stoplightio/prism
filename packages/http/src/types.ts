import { IPrism, IPrismComponents, IPrismConfig } from '@stoplight/prism-core';
import { IHttpOperation, Omit } from '@stoplight/types';

export type TPrismHttpInstance<LoaderInput> = IPrism<
  IHttpOperation,
  IHttpRequest,
  IHttpResponse,
  IHttpConfig,
  LoaderInput
>;

export type TPrismHttpComponents<LoaderInput> = Partial<
  IPrismComponents<IHttpOperation, IHttpRequest, IHttpResponse, IHttpConfig, LoaderInput>
>;

// TODO: should be complete | and in the @stoplight/types repo
export type IHttpMethod = 'get' | 'put' | 'post' | 'delete' | 'options' | 'head' | 'patch' | 'trace'; // ... etc

export interface IHttpOperationConfig {
  mediaType?: string;
  code?: string;
  exampleKey?: string;
  dynamic: boolean;
}

export interface IHttpConfig extends IPrismConfig {
  mock: false | IHttpOperationConfig;

  validate?: {
    request?:
      | boolean
      | {
          hijack?: boolean;
          headers?: boolean;
          query?: boolean;
          body?: boolean;
        };

    response?:
      | boolean
      | {
          headers?: boolean;
          body?: boolean;
        };
  };
}

export interface IHttpNameValues {
  [name: string]: string | string[];
}

export interface IHttpNameValue {
  [name: string]: string;
}

export interface IHttpUrl {
  baseUrl?: string;
  path: string;
  query?: IHttpNameValues;
}

export interface IHttpRequest {
  method: IHttpMethod;
  url: IHttpUrl;
  headers?: IHttpNameValue;
  body?: any;
}

export interface IHttpResponse {
  statusCode: number;
  headers?: IHttpNameValue;
  body?: any;
}

export type ProblemJson = {
  type: string;
  title: string;
  status: number;
  detail: string;
};

export class ProblemJsonError extends Error {
  public static fromTemplate(template: Omit<ProblemJson, 'detail'>, detail?: string): ProblemJsonError {
    Error.captureStackTrace(this, ProblemJsonError);
    return new ProblemJsonError(
      `https://stoplight.io/prism/errors#${template.type}`,
      template.title,
      template.status,
      detail || '',
    );
  }

  public static fromPlainError(error: Error & { detail?: string; status?: number }): ProblemJson {
    return {
      type: error.name && error.name !== 'Error' ? error.name : 'https://stoplight.io/prism/errors#UNKNOWN',
      title: error.message,
      status: error.status || 500,
      detail: error.detail || '',
    };
  }

  constructor(readonly name: string, readonly message: string, readonly status: number, readonly detail: string) {
    super(message);
    Error.captureStackTrace(this, ProblemJsonError);
  }
}

export type PayloadGenerator = (f: unknown) => Promise<unknown>;
