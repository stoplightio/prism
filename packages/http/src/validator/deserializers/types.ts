import { HttpParamStyles } from '@stoplight/types';
import { IHttpNameValue, IHttpNameValues, JSONSchema } from '../../types';

export type deserializeFn<Parameters> = (
  name: string,
  parameters: Parameters,
  schema?: JSONSchema,
  explode?: boolean
) => unknown;

export interface IHttpParamDeserializerRegistry<Parameters, S = HttpParamStyles> {
  get(style: S): IHttpParamStyleDeserializer<Parameters> | undefined;
}

export interface IHttpParamStyleDeserializer<Parameters, S = HttpParamStyles> {
  supports: (style: S) => boolean;
  deserialize: deserializeFn<Parameters>;
}

export type IHttpHeaderParamStyleDeserializer = IHttpParamStyleDeserializer<IHttpNameValue>;
export type IHttpQueryParamStyleDeserializer = IHttpParamStyleDeserializer<IHttpNameValues>;
