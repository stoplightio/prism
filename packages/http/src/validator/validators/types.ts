import { IPrismDiagnostic } from '@stoplight/prism-core/src/types';
import { JSONSchema } from 'http/src/types';

export interface IHttpValidator<Target, Specs> {
  validate(target: Target, specs: Specs[], mediaType?: string): IPrismDiagnostic[];
}

export interface ISchemaValidator<S = JSONSchema> {
  validate(content: any, schema: S): IPrismDiagnostic[];
  supports(mediaType: string): boolean;
}

export interface IValidatorRegistry {
  get(mediaType: string): ((content: any, schema: JSONSchema) => IPrismDiagnostic[]) | undefined;
}
