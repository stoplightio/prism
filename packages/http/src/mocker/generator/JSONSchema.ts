// @ts-ignore
import * as jsf from '@stoplight/json-schema-faker';
import { cloneDeep } from 'lodash';

jsf.option({
  failOnInvalidTypes: false,
  failOnInvalidFormat: false,
  alwaysFakeOptionals: true,
  optionalsProbability: 1,
  fixedProbabilities: true,
  ignoreMissingRefs: true,
  useExamplesValue: true,
  useDefaultValue: true,
  maxItems: 20,
  maxLength: 100,
});

export async function generate(source: unknown): Promise<unknown> {
  return jsf.resolve(cloneDeep(source));
}
