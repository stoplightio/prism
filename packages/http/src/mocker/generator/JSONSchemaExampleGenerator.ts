// @ts-ignore
import * as jsf from '@stoplight/json-schema-faker';
import { cloneDeep } from 'lodash';
import { IExampleGenerator } from './IExampleGenerator';

jsf.option({
  failOnInvalidTypes: false,
  failOnInvalidFormat: false,
  alwaysFakeOptionals: true,
  optionalsProbability: 1,
  fixedProbabilities: true,
  ignoreMissingRefs: true,
  useExamplesValue: true,
  useDefaultValue: true,
  maxItems: 20,
  maxLength: 100,
});

export class JSONSchemaExampleGenerator implements IExampleGenerator {
  public async generate(schema: unknown, mediaType: string): Promise<string> {
    const example = await jsf.resolve(cloneDeep(schema));
    return this.transform(mediaType, example);
  }

  private transform(mediaType: string, input: object): string {
    switch (mediaType) {
      case 'application/json':
        return JSON.stringify(input);

      default:
        throw new Error(`Unknown media type '${mediaType}'`);
    }
  }
}
