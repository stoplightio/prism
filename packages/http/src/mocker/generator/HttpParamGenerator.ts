import { IHttpContent, IHttpParam, INodeExample, INodeExternalExample } from '@stoplight/types';
import * as O from 'fp-ts/lib/Option';
import { pipe } from 'fp-ts/lib/pipeable';
import { JSONSchema } from '../../types';
import { generate as generateDynamicExample } from './JSONSchema';

export function improveSchema(schema: JSONSchema): JSONSchema {
  const newSchema = { ...schema };

  if (newSchema.type === 'integer' || newSchema.type === 'number') {
    if (!newSchema.minimum) {
      newSchema.minimum = 1;
    }

    if (!newSchema.maximum) {
      newSchema.maximum = 1000;
    }
  }

  if (newSchema.type === 'string') {
    if (!newSchema.format && !newSchema.enum && !newSchema.pattern) {
      newSchema['x-faker'] = 'lorem.word';
    }
  }

  if (newSchema.type === 'object') {
    if (newSchema.properties) {
      newSchema.properties = Object.entries(newSchema.properties).reduce((r, [k, v]) => {
        r[k] = improveSchema(v);
        return r;
      }, {});
    }
  }

  if (newSchema.type === 'array') {
    if (typeof newSchema.items === 'object') {
      newSchema.items = improveSchema(newSchema.items);
    }
  }

  return newSchema;
}

function pickStaticExample(examples: O.Option<Array<INodeExample | INodeExternalExample>>): O.Option<unknown> {
  return pipe(
    examples,
    O.mapNullable(exs => exs[Math.floor(Math.random() * exs.length)]),
    O.mapNullable(example => (example as INodeExample).value)
  );
}

export function generate(param: IHttpParam | IHttpContent): O.Option<unknown> {
  return pipe(
    O.fromNullable(param.examples),
    pickStaticExample,
    O.alt(() =>
      pipe(
        O.fromNullable(param.schema),
        O.map(improveSchema),
        O.chain(schema => O.fromEither(generateDynamicExample(schema)))
      )
    )
  );
}
