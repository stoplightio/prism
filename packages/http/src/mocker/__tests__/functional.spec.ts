import { ISchema } from '@stoplight/types';
import * as Ajv from 'ajv';

import { httpOperations, httpRequests } from '../../__tests__/fixtures';
import { JSONSchemaExampleGenerator } from '../generator/JSONSchemaExampleGenerator';
import { HttpMocker } from '../index';

describe('http mocker', () => {
  const mocker = new HttpMocker(new JSONSchemaExampleGenerator());

  describe('request is valid', () => {
    describe('given only enforced content type', () => {
      test('and that content type exists should first 200 static example', async () => {
        const response = await mocker.mock({
          resource: httpOperations[0],
          input: httpRequests[0],
          config: {
            mock: {
              dynamic: false,
              mediaType: 'text/plain',
            },
          },
        });

        expect(response).toMatchSnapshot();
      });

      test('and that content type does not exist should return empty body', () => {
        return expect(
          mocker.mock({
            resource: httpOperations[0],
            input: httpRequests[0],
            config: {
              mock: {
                dynamic: false,
                mediaType: 'text/funky',
              },
            },
          }),
        ).resolves.toMatchObject({ headers: { 'Content-type': 'text/plain' }, body: undefined });
      });
    });

    describe('given enforced status code and contentType and exampleKey', () => {
      test('should return the matching example', async () => {
        const response = await mocker.mock({
          resource: httpOperations[0],
          input: httpRequests[0],
          config: {
            mock: {
              dynamic: false,
              code: '201',
              exampleKey: 'second',
              mediaType: 'application/xml',
            },
          },
        });

        expect(response).toMatchSnapshot();
      });
    });

    describe('given enforced status code and contentType', () => {
      test('should return the first matching example', async () => {
        const response = await mocker.mock({
          resource: httpOperations[0],
          input: httpRequests[0],
          config: {
            mock: {
              dynamic: false,
              code: '201',
              mediaType: 'application/xml',
            },
          },
        });

        expect(response).toMatchSnapshot();
      });
    });

    describe('given enforced example key', () => {
      test('should return application/json, 200 response', async () => {
        const response = await mocker.mock({
          resource: httpOperations[0],
          input: httpRequests[0],
          config: {
            mock: {
              dynamic: false,
              exampleKey: 'bear',
            },
          },
        });

        expect(response).toMatchSnapshot();
      });

      test('and mediaType should return 200 response', async () => {
        const response = await mocker.mock({
          resource: httpOperations[0],
          input: httpRequests[0],
          config: {
            mock: {
              dynamic: false,
              exampleKey: 'second',
              mediaType: 'application/xml',
            },
          },
        });

        expect(response).toMatchSnapshot();
      });
    });

    describe('given enforced status code', () => {
      test('should return the first matching example of application/json', async () => {
        const response = await mocker.mock({
          resource: httpOperations[0],
          input: httpRequests[0],
          config: {
            mock: {
              dynamic: false,
              code: '201',
            },
          },
        });

        expect(response).toMatchSnapshot();
      });

      test('given that status code is not defined should throw an error', () => {
        const rejection = mocker.mock({
          resource: httpOperations[0],
          input: httpRequests[0],
          config: {
            mock: {
              dynamic: false,
              code: '205',
            },
          },
        });

        return expect(rejection).rejects.toEqual(new Error('Requested status code is not defined in the schema.'));
      });

      test('and example key should return application/json example', async () => {
        const response = await mocker.mock({
          resource: httpOperations[0],
          input: httpRequests[0],
          config: {
            mock: {
              dynamic: false,
              code: '201',
              exampleKey: 'second',
            },
          },
        });

        expect(response).toMatchSnapshot();
      });
    });

    describe('HttpOperation contains example', () => {
      test('return lowest 2xx code and match response example to media type accepted by request', async () => {
        const response = await mocker.mock({
          resource: httpOperations[0],
          input: httpRequests[0],
        });

        expect(response).toMatchSnapshot();
      });

      test('return lowest 2xx response and the first example matching the media type', async () => {
        const response = await mocker.mock({
          resource: httpOperations[1],
          input: Object.assign({}, httpRequests[0], {
            data: Object.assign({}, httpRequests[0].data, {
              headers: { 'Content-type': 'application/xml' },
            }),
          }),
        });

        expect(response).toMatchSnapshot();
      });

      describe('the media type requested does not match the example', () => {
        test('throw exception', () => {
          return expect(
            mocker.mock({
              resource: httpOperations[0],
              input: Object.assign({}, httpRequests[0], {
                data: Object.assign({}, httpRequests[0].data, {
                  headers: { 'Content-type': 'application/yaml' },
                }),
              }),
            }),
          ).resolves.toMatchObject({
            headers: { 'Content-type': 'text/plain' },
            body: undefined,
          });
        });
      });
    });

    describe('HTTPOperation contain no examples', () => {
      test('return dynamic response', async () => {
        if (!httpOperations[1].responses[0].contents[0].schema) {
          throw new Error('Missing test');
        }

        const ajv = new Ajv();
        const validate = ajv.compile(httpOperations[1].responses[0].contents[0].schema as ISchema);

        const response = await mocker.mock({
          resource: httpOperations[1],
          input: httpRequests[0],
          config: {
            mock: {
              dynamic: true,
            },
          },
        });

        expect(response).toHaveProperty('statusCode', 200);
        expect(response).toHaveProperty('headers', { 'Content-type': 'application/json' });
        expect(validate(JSON.parse(response.body))).toBe(true);
      });
    });
  });

  describe('request is invalid', () => {
    test('returns 422 and static error response', async () => {
      const response = await mocker.mock({
        resource: httpOperations[0],
        input: httpRequests[1],
      });

      expect(response).toMatchSnapshot();
    });

    test('returns 422 and dynamic error response', async () => {
      if (!httpOperations[1].responses[1].contents[0].schema) {
        throw new Error('Missing test');
      }

      const response = await mocker.mock({
        resource: httpOperations[1],
        input: httpRequests[1],
      });

      const ajv = new Ajv();
      const validate = ajv.compile(httpOperations[1].responses[1].contents[0].schema as ISchema);

      expect(validate(JSON.parse(response.body))).toBe(true);
    });
  });
});
