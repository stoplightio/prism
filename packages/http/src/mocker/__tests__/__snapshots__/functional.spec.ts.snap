// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`http mocker request is valid HttpOperation contains example return lowest 2xx response and the first example matching the media type 1`] = `
Object {
  "body": "<todo><name>Shopping</name><completed>false</completed></todo>",
  "headers": Object {
    "Content-type": "application/xml",
    "x-todos-publish": Any<String>,
  },
  "statusCode": 200,
}
`;

exports[`http mocker request is valid given enforced example key and mediaType should return 200 response 1`] = `
Object {
  "body": "{ \\"root\\": \\"second\\" }",
  "headers": Object {
    "Content-type": "application/xml",
  },
  "statusCode": 200,
}
`;

exports[`http mocker request is valid given enforced example key should return application/json, 200 response 1`] = `
Object {
  "body": Array [
    Object {
      "completed": false,
      "id": 2,
      "name": "make bears",
    },
  ],
  "headers": Object {
    "Content-type": "application/json",
  },
  "statusCode": 200,
}
`;

exports[`http mocker request is valid given enforced status code and contentType and exampleKey should return the matching example 1`] = `
Object {
  "body": "<root>second</root>",
  "headers": Object {
    "Content-type": "application/xml",
  },
  "statusCode": 201,
}
`;

exports[`http mocker request is valid given enforced status code and contentType should return the first matching example 1`] = `
Object {
  "body": "<root>first</root>",
  "headers": Object {
    "Content-type": "application/xml",
  },
  "statusCode": 201,
}
`;

exports[`http mocker request is valid given enforced status code and example key should return application/json example 1`] = `
Object {
  "body": "{ \\"root\\": \\"second\\" }",
  "headers": Object {
    "Content-type": "application/json",
  },
  "statusCode": 201,
}
`;

exports[`http mocker request is valid given enforced status code should return the first matching example of application/json 1`] = `
Object {
  "body": "{ \\"root\\": \\"first\\" }",
  "headers": Object {
    "Content-type": "application/json",
  },
  "statusCode": 201,
}
`;

exports[`http mocker request is valid given only enforced content type and that content type exists should first 200 static example 1`] = `
Object {
  "body": "some text",
  "headers": Object {
    "Content-type": "text/plain",
  },
  "statusCode": 200,
}
`;
